src/spreadsheets.js

```diff
@@ -0,0 +1,286 @@
+var config = require('../config.json');
+var fluid = fluid || require('infusion');
+
+var CREDENTIALS_PATH = config.credentialsPath;
+var TOKEN_PATH = config.tokenPath;
+var SCOPES = config.scopes;
+
+/**********************
+ * SPREADSHEET GRADES *
+ **********************/
+
+var spreadsheets = fluid.registerNamespace('spreadsheets');
+module.exports = spreadsheets;
+
+// This logging function is used to trigger a breakpoint and get component tree state
+spreadsheets.debug = function(...messages) {
+    console.log(messages);
+}
+
+/**
+ * component for reading and writing entire spreadsheets, located based on their id/URL
+ */
+fluid.defaults('spreadsheets.spreadsheet', {
+    gradeNames: ['fluid.modelComponent'],
+    spreadsheetId: '', // the id of the host spreadsheet, which can be extracted from the URL
+    members: {
+        sheetBounds: {} // the coordinate ranges containing all data in each sheet 'tab', keyed by the sheetName
+        // this may be significantly larger than the used cells without incurring any computational cost
+    },
+    components: {
+        apiClient: { // the authorized API client
+            type: 'spreadsheets.sheetsAPIClient'
+        }
+    },
+    // TODO: ask A about old-style event passing versus current setup, 
+    // which is an onCreate listener triggering an invoker 
+    // '{that}.apiClient.events.onResolve': 'spreadsheets.cell.updateValue',
+    // '{that}.apiClient.events.onError': 'spreadsheets.cell.displayError'
+    listeners: { // read the remote spreadsheet on component creation
+        'onCreate.getRemoteContent': 'spreadsheets.spreadsheet.getRemoteContent({that}, {apiClient})'
+    },
+    // When the model is written to, update the remote sheet
+    // TODO: where is the change source stored?
+    modelListeners: {
+        '*': {
+            func: 'spreadsheets.spreadsheet.setRemoteContent',
+            args: ['{that}', '{apiClient}', '{change}.path.0', '{change}.value'],
+            // FIXME: currently, this model relay doesn't work as intended. 
+            // It should systematically exclude mutations coming from reads. 
+            // Can those be associated with a particular invoker on the apiClient sub-component or locally? 
+            // The kludge I can make right now is to check the written value against the last-read value.
+            // excludeSource: 'init' does not seem to be what I want, because multiple reads could happen over the component life cycle.
+        }
+    }
+});
+
+// TODO: refactor this messy function
+// TODO: I wonder if I can more declaratively specify
+//   which api call with which body
+//   how to decode and extract data
+//   how to fill the model
+// and whether that has value for the design
+spreadsheets.spreadsheet.getRemoteContent = function(that, apiClient) {
+    var metadataPromise = apiClient.getSpreadsheet(apiClient, that.options.spreadsheetId);
+    var requests = [];
+    metadataPromise.then(function(res) {
```

Comment:
As we discussed, this could be broken up somewhat by being rexpressed as a transforming promise chain.